#  EDUNET FOUNDATION-Class Exercise Notebook

# Lab 31 Implement Hypertuning Techniques

Hypertuning techniques focus on identifying optimal values for Hyperparameters

Identifying hyperparameters which may affect model performance

Generally, hyperparameters in a deep neural network are: 
- Number of iterations 
- Number of layers in the neural network 
- Number of hidden units in each layer 
- Learning rate α 
- Step size 
- Choice of the activation function 
- Losss function 
- Mini-batch Size 
- Momentum 
- Regularization 
- Drop out rate 
- Weight Decay

For the practice, first we need an AI model to hypertune.

So, let's develop a DL model. Dataset, used here, can be downloaded from this link: https://www.kaggle.com/datasets/karakaggle/kaggle-cat-vs-dog-dataset

In [1]:
# Importing the Keras libraries and packages
from keras.models import Sequential
from keras.layers import Convolution2D
from keras.layers import MaxPooling2D
from keras.layers import Flatten
from keras.layers import Dense

In [2]:
#Initialising the CNN
classifier = Sequential()

# Step 1 - Convolution
classifier.add(Convolution2D(32, 3, 3, input_shape = (64,64, 3), activation = 'relu'))

# Step 2 - Pooling
classifier.add(MaxPooling2D(pool_size = (2, 2)))

# Adding a second convolutional layer
classifier.add(Convolution2D(32, 3, 3, activation = 'relu'))
classifier.add(MaxPooling2D(pool_size = (2, 2)))

# Step 3 - Flattening
classifier.add(Flatten())

# Step 4 - Full connection
classifier.add(Dense(128, activation = 'relu'))
classifier.add(Dense(1, activation = 'sigmoid'))

# Compiling the CNN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [4]:
from keras.preprocessing.image import ImageDataGenerator

# Data Augmentation

train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   zoom_range = 0.2,
                                   horizontal_flip = True)

test_datagen = ImageDataGenerator(rescale = 1./255)

training_set = train_datagen.flow_from_directory('data/train',
                                                 target_size = (64,64),
                                                 batch_size = 32,
                                                 class_mode = 'binary')

test_set = test_datagen.flow_from_directory('data/test',
                                            target_size = (64,64),
                                            batch_size = 32,
                                            class_mode = 'binary')

Found 165 images belonging to 2 classes.
Found 30 images belonging to 2 classes.


In [5]:
classifier.fit(training_set,# the training set
                         epochs = 10,# number of epochs 
                         validation_data = test_set# the test set
                         )

Epoch 1/10
6/6 [==============================] - 8s 946ms/step - loss: 0.6948 - accuracy: 0.5152 - val_loss: 0.6974 - val_accuracy: 0.5000
Epoch 2/10
6/6 [==============================] - 2s 363ms/step - loss: 0.6912 - accuracy: 0.5152 - val_loss: 0.6938 - val_accuracy: 0.5000
Epoch 3/10
6/6 [==============================] - 3s 376ms/step - loss: 0.6918 - accuracy: 0.5394 - val_loss: 0.6947 - val_accuracy: 0.5000
Epoch 4/10
6/6 [==============================] - 2s 393ms/step - loss: 0.6907 - accuracy: 0.5030 - val_loss: 0.6940 - val_accuracy: 0.5000
Epoch 5/10
6/6 [==============================] - 2s 355ms/step - loss: 0.6904 - accuracy: 0.5212 - val_loss: 0.6941 - val_accuracy: 0.5000
Epoch 6/10
6/6 [==============================] - 3s 502ms/step - loss: 0.6885 - accuracy: 0.5455 - val_loss: 0.7010 - val_accuracy: 0.5000
Epoch 7/10
6/6 [==============================] - 2s 406ms/step - loss: 0.6923 - accuracy: 0.5152 - val_loss: 0.7020 - val_accuracy: 0.5000
Epoch 8/10
6/6 [====

Final validation accuracy is 0.5

In the above implemtation, we will change some of the hyperparameters to find best suited set of Hyperparameters 

Variation 1: Changing the number of epochs

In [6]:
classifier.fit(training_set,# the training set
                         epochs = 50,# number of epochs 
                         validation_data = test_set# the test set
                         )

Epoch 1/50
6/6 [==============================] - 2s 398ms/step - loss: 0.6832 - accuracy: 0.6000 - val_loss: 0.6925 - val_accuracy: 0.6000
Epoch 2/50
6/6 [==============================] - 2s 318ms/step - loss: 0.6763 - accuracy: 0.6424 - val_loss: 0.6919 - val_accuracy: 0.6333
Epoch 3/50
6/6 [==============================] - 2s 428ms/step - loss: 0.6688 - accuracy: 0.6303 - val_loss: 0.6938 - val_accuracy: 0.6333
Epoch 4/50
6/6 [==============================] - 2s 404ms/step - loss: 0.6641 - accuracy: 0.6545 - val_loss: 0.6962 - val_accuracy: 0.5667
Epoch 5/50
6/6 [==============================] - 2s 330ms/step - loss: 0.6505 - accuracy: 0.6424 - val_loss: 0.7045 - val_accuracy: 0.5667
Epoch 6/50
6/6 [==============================] - 2s 283ms/step - loss: 0.6581 - accuracy: 0.5939 - val_loss: 0.7131 - val_accuracy: 0.4333
Epoch 7/50
6/6 [==============================] - 3s 457ms/step - loss: 0.6567 - accuracy: 0.5576 - val_loss: 0.7180 - val_accuracy: 0.5000
Epoch 8/50
6/6 [====

Final validation accuracy is now increased to 0.6333

Variation 2: Changing the number of units in the model architecture but keeping the epoch same as 10

In [8]:
#Initialising the CNN
classifier2 = Sequential()

# Step 1 - Convolution
classifier2.add(Convolution2D(64, 3, 3, input_shape = (64,64, 3), activation = 'relu'))

# Step 2 - Pooling
classifier2.add(MaxPooling2D(pool_size = (2, 2)))

# Adding a second convolutional layer
classifier2.add(Convolution2D(64, 3, 3, activation = 'relu'))
classifier2.add(MaxPooling2D(pool_size = (2, 2)))

# Step 3 - Flattening
classifier2.add(Flatten())

# Step 4 - Full connection
classifier2.add(Dense(256, activation = 'relu'))
classifier2.add(Dense(1, activation = 'sigmoid'))

# Compiling the CNN
classifier2.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [10]:
classifier2.fit(training_set,# the training set
                         epochs = 10,# number of epochs 
                         validation_data = test_set# the test set
                         )

Epoch 1/10
6/6 [==============================] - 22s 2s/step - loss: 0.6960 - accuracy: 0.4606 - val_loss: 0.6970 - val_accuracy: 0.5000
Epoch 2/10
6/6 [==============================] - 10s 2s/step - loss: 0.6944 - accuracy: 0.4909 - val_loss: 0.6943 - val_accuracy: 0.5000
Epoch 3/10
6/6 [==============================] - 10s 2s/step - loss: 0.6889 - accuracy: 0.5333 - val_loss: 0.6994 - val_accuracy: 0.5000
Epoch 4/10
6/6 [==============================] - 11s 2s/step - loss: 0.6889 - accuracy: 0.5758 - val_loss: 0.6924 - val_accuracy: 0.5667
Epoch 5/10
6/6 [==============================] - 10s 2s/step - loss: 0.6885 - accuracy: 0.5939 - val_loss: 0.6923 - val_accuracy: 0.5667
Epoch 6/10
6/6 [==============================] - 10s 2s/step - loss: 0.6895 - accuracy: 0.5152 - val_loss: 0.6977 - val_accuracy: 0.5000
Epoch 7/10
6/6 [==============================] - 8s 1s/step - loss: 0.6990 - accuracy: 0.4848 - val_loss: 0.6961 - val_accuracy: 0.5000
Epoch 8/10
6/6 [===================

Final validation accuracy remained unaffected.

Variation 3: Changing optimizer 

In [11]:
#Initialising the CNN
classifier3 = Sequential()

# Step 1 - Convolution
classifier3.add(Convolution2D(64, 3, 3, input_shape = (64,64, 3), activation = 'relu'))

# Step 2 - Pooling
classifier3.add(MaxPooling2D(pool_size = (2, 2)))

# Adding a second convolutional layer
classifier3.add(Convolution2D(64, 3, 3, activation = 'relu'))
classifier3.add(MaxPooling2D(pool_size = (2, 2)))

# Step 3 - Flattening
classifier3.add(Flatten())

# Step 4 - Full connection
classifier3.add(Dense(256, activation = 'relu'))
classifier3.add(Dense(1, activation = 'sigmoid'))

# Compiling the CNN
classifier3.compile(optimizer = 'SGD', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [12]:
classifier3.fit(training_set,# the training set
                         epochs = 10,# number of epochs 
                         validation_data = test_set# the test set
                         )

Epoch 1/10
6/6 [==============================] - 2s 202ms/step - loss: 0.6935 - accuracy: 0.4848 - val_loss: 0.6913 - val_accuracy: 0.5000
Epoch 2/10
6/6 [==============================] - 1s 171ms/step - loss: 0.6932 - accuracy: 0.4848 - val_loss: 0.6914 - val_accuracy: 0.5000
Epoch 3/10
6/6 [==============================] - 1s 145ms/step - loss: 0.6928 - accuracy: 0.4909 - val_loss: 0.6912 - val_accuracy: 0.5000
Epoch 4/10
6/6 [==============================] - 1s 157ms/step - loss: 0.6933 - accuracy: 0.4788 - val_loss: 0.6913 - val_accuracy: 0.5000
Epoch 5/10
6/6 [==============================] - 1s 152ms/step - loss: 0.6938 - accuracy: 0.4909 - val_loss: 0.6911 - val_accuracy: 0.5000
Epoch 6/10
6/6 [==============================] - 1s 162ms/step - loss: 0.6943 - accuracy: 0.4727 - val_loss: 0.6915 - val_accuracy: 0.5000
Epoch 7/10
6/6 [==============================] - 1s 152ms/step - loss: 0.6931 - accuracy: 0.4970 - val_loss: 0.6916 - val_accuracy: 0.4667
Epoch 8/10
6/6 [====

Accuracy, slightly, improved to 0.533

In this way, you can hypertune your model.
Remember, there is no definite set of hyperparameters which will work in every scenario.